controllers/goalertintegration/goalertintegration_controller.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package goalertintegration
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/go-logr/logr"
+	goalertv1alpha1 "github.com/openshift/configure-goalert-operator/api/v1alpha1"
+	"github.com/openshift/configure-goalert-operator/pkg/utils"
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	GoalertFinalizerPrefix = "goalert.managed.openshift.io/goalert-"
+)
+
+// GoalertIntegrationReconciler reconciles a GoalertIntegration object
+type GoalertIntegrationReconciler struct {
+	client.Client
+	Scheme    *runtime.Scheme
+	reqLogger logr.Logger
+}
+
+//+kubebuilder:rbac:groups=goalert.managed.openshift.io,resources=goalertintegrations,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=goalert.managed.openshift.io,resources=goalertintegrations/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=goalert.managed.openshift.io,resources=goalertintegrations/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the GoalertIntegration object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *GoalertIntegrationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	// TODO(user): your logic here
+	gi := &goalertv1alpha1.GoalertIntegration{}
+	err := r.Get(context.TODO(), req.NamespacedName, gi)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return r.doNotRequeue()
+		}
+		// Error reading the object - requeue the request.
+		return r.requeueOnErr(err)
+	}
+
```

Comment:
Typically the order for each reconcile is:

1. Get the relevant CR
2. Check if it's deleting and cleanup if so
3. Ensure state if it's not deleting

atm 2 and 3 are flipped - the benefit with switching this around is that the controller doesn't spend time ensuring state each run if it's just going to start deleting in the end. I recommend moving the deletion check up here.